File: src/algorithms/arrays/dutch-national-flag/sources/dutch-national-flag.ts
// Dutch National Flag — O(n) 3-way partition using three pointers (low, mid, high)
function dutchNationalFlag(inputArray: number[]): number[] {
  const result = [...inputArray];
  let lowPointer = 0; // @step:initialize
  let midPointer = 0; // @step:initialize
  let highPointer = result.length - 1; // @step:initialize

  while (midPointer <= highPointer) {
    const currentValue = result[midPointer]!; // @step:compare

    if (currentValue === 0) {
      // @step:compare
      const tempValue = result[lowPointer]!; // @step:swap
      result[lowPointer] = result[midPointer]!; // @step:swap
      result[midPointer] = tempValue; // @step:swap
      lowPointer++; // @step:visit
      midPointer++; // @step:visit
    } else if (currentValue === 1) {
      // @step:compare
      midPointer++; // @step:visit
    } else {
      const tempValue = result[highPointer]!; // @step:swap
      result[highPointer] = result[midPointer]!; // @step:swap
      result[midPointer] = tempValue; // @step:swap
      highPointer--; // @step:visit
    }
  }

  return result; // @step:complete
}

File: src/algorithms/arrays/sorting-partitioning/dutch-national-flag/sources/dutch-national-flag.rs
// Dutch National Flag — O(n) 3-way partition using three pointers (low, mid, high)
fn dutch_national_flag(input_array: &[i32]) -> Vec<i32> {
    let mut result = input_array.to_vec();
    let mut low_pointer = 0usize; // @step:initialize
    let mut mid_pointer = 0usize; // @step:initialize
    let mut high_pointer = if result.is_empty() { 0 } else { result.len() - 1 }; // @step:initialize

    while mid_pointer <= high_pointer && !result.is_empty() {
        let current_value = result[mid_pointer]; // @step:compare

        if current_value == 0 {
            // @step:compare
            result.swap(low_pointer, mid_pointer); // @step:swap
            low_pointer += 1; // @step:visit
            mid_pointer += 1; // @step:visit
        } else if current_value == 1 {
            // @step:compare
            mid_pointer += 1; // @step:visit
        } else {
            result.swap(mid_pointer, high_pointer); // @step:swap
            if high_pointer == 0 { break; }
            high_pointer -= 1; // @step:visit
        }
    }

    result // @step:complete
}
